File: wrappers/dotnet/indy-sdk-dotnet-test/Wrapper/AgentTests/AgentSendTest.cs
using Indy.Sdk.Dotnet.Wrapper;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System;
using System.Threading.Tasks;
using static Indy.Sdk.Dotnet.Wrapper.Agent;

namespace Indy.Sdk.Dotnet.Test.Wrapper.AgentTests
{
    [TestClass]
    public class AgentSendTest : AgentIntegrationTestBase
    {
        private static TaskCompletionSource<Connection> serverToClientConnectionFuture = new TaskCompletionSource<Connection>();
        private static TaskCompletionSource<string> serverToClientMsgFuture = new TaskCompletionSource<string>();
        private static TaskCompletionSource<string> clientToServerMsgFuture = new TaskCompletionSource<string>();

        private new static MessageReceivedHandler _messageObserver = (Connection connection, string message) =>
            {
                Console.WriteLine("Received message '" + message + "' on connection " + connection);

                serverToClientMsgFuture.SetResult(message);
            };

        private new static MessageReceivedHandler _messageObserverForIncoming = (Connection connection, string message) =>
            {
                Console.WriteLine("Received message '" + message + "' on incoming connection " + connection);

                clientToServerMsgFuture.SetResult(message);
            };

        private new static ConnectionOpenedHandler _incomingConnectionObserver = (Listener listener, Connection connection, string senderDid, string receiverDid) =>
            {
                Console.WriteLine("New connection " + connection);

                serverToClientConnectionFuture.SetResult(connection);

                return _messageObserverForIncoming;
            };

        [TestMethod]
        public async Task TestAgentSendWorks()
        {
            var endpoint = "127.0.0.1:9609";

            var myDidResult = await Signus.CreateAndStoreMyDidAsync(_wallet, "{}");

            var identityJson = string.Format("{{\"did\":\"{0}\", \"pk\":\"{1}\", \"verkey\":\"{2}\", \"endpoint\":\"{3}\"}}",
                    myDidResult.Did, myDidResult.Pk, myDidResult.VerKey, endpoint);
            await Signus.StoreTheirDidAsync(_wallet, identityJson);

            var activeListener = await Agent.AgentListenAsync(endpoint, _incomingConnectionObserver);

            await activeListener.AddIdentityAsync(_pool, _wallet, myDidResult.Did);

            var clientToServerConnection = await Agent.AgentConnectAsync(_pool, _wallet, myDidResult.Did, myDidResult.Did, _messageObserver);

            var clientToServerMessage = "msg_from_client";
            var serverToClientMessage = "msg_from_server";

            await clientToServerConnection.SendAsync(clientToServerMessage);

            Assert.AreEqual(clientToServerMessage, await clientToServerMsgFuture.Task);

            var serverToClientConnection = await serverToClientConnectionFuture.Task;
            await serverToClientConnection.SendAsync(serverToClientMessage);

            Assert.AreEqual(serverToClientMessage, await serverToClientMsgFuture.Task);
        }
    }
}

File: wrappers/java/src/test/java/org/hyperledger/indy/sdk/agent/AgentSendTest.java
package org.hyperledger.indy.sdk.agent;

import org.hyperledger.indy.sdk.agent.Agent.Connection;
import org.hyperledger.indy.sdk.agent.Agent.Listener;
import org.hyperledger.indy.sdk.signus.Signus;
import org.hyperledger.indy.sdk.signus.SignusResults;
import org.junit.Test;

import java.util.concurrent.CompletableFuture;

import static org.junit.Assert.assertEquals;

public class AgentSendTest extends AgentIntegrationTest {

	private static CompletableFuture<Connection> serverToClientConnectionFuture = new CompletableFuture<Connection>();
	private static CompletableFuture<String> serverToClientMsgFuture = new CompletableFuture<String>();
	private static CompletableFuture<String> clientToServerMsgFuture = new CompletableFuture<String>();

	private static final AgentObservers.MessageObserver messageObserver = new AgentObservers.MessageObserver() {

		public void onMessage(Connection connection, String message) {

			System.out.println("Received message '" + message + "' on connection " + connection);

			serverToClientMsgFuture.complete(message);
		}
	};

	private static final AgentObservers.MessageObserver messageObserverForIncoming = new AgentObservers.MessageObserver() {

		public void onMessage(Connection connection, String message) {

			System.out.println("Received message '" + message + "' on incoming connection " + connection);

			clientToServerMsgFuture.complete(message);
		}
	};

	private static final AgentObservers.ConnectionObserver incomingConnectionObserver = new AgentObservers.ConnectionObserver() {

		public AgentObservers.MessageObserver onConnection(Listener listener, Connection connection, String senderDid, String receiverDid) {

			System.out.println("New connection " + connection);

			serverToClientConnectionFuture.complete(connection);

			return messageObserverForIncoming;
		}
	};

	@Test
	public void testAgentSendWorksForAllDataInWalletPresent() throws Exception {
		String endpoint = "127.0.0.1:9609";

		SignusResults.CreateAndStoreMyDidResult myDid = Signus.createAndStoreMyDid(wallet, "{}").get();

		String identityJson = String.format(AGENT_IDENTITY_JSON_TEMPLATE, myDid.getDid(), myDid.getPk(), myDid.getVerkey(), endpoint);
		Signus.storeTheirDid(wallet, identityJson).get();

		Listener activeListener = Agent.agentListen(endpoint, incomingConnectionObserver).get();

		activeListener.agentAddIdentity(pool, wallet, myDid.getDid()).get();

		Connection clientToServerConnection = Agent.agentConnect(pool, wallet, myDid.getDid(), myDid.getDid(), messageObserver).get();

		String clientToServerMessage = "msg_from_client";
		String serverToClientMessage = "msg_from_server";

		clientToServerConnection.agentSend(clientToServerMessage).get();

		assertEquals(clientToServerMessage, clientToServerMsgFuture.get());

		Connection serverToClientConnection = serverToClientConnectionFuture.get();
		serverToClientConnection.agentSend(serverToClientMessage).get();

		assertEquals(serverToClientMessage, serverToClientMsgFuture.get());
	}
}